lib/compiler/src/compile_messages.erl

```diff
@@ -0,0 +1,214 @@
+    %% this is the location format used in the type analysis pass
+    Src = quote_source(F, StartLoc, EndLoc, Opts),
+    io:fwrite("~ts:~ts: ~ts\n~ts", [F, fmt_pos(StartLoc), Mod:format_error(E), Src]),
+    list_errors(F, Es, Opts);
+list_errors(F, [{Loc, Mod, E} | Es], Opts) ->
+    StartLoc = erl_anno:location(Loc),
+    EndLoc =
+        case erl_anno:end_location(Loc) of
+            undefined -> StartLoc;
+            Loc2 -> Loc2
+        end,
+    list_errors(F, [{{StartLoc, EndLoc}, Mod, E} | Es], Opts);
+list_errors(_F, [], _Opts) ->
+    ok.
+
+fmt_pos({Line, Col}) ->
+    io_lib:format("~w:~w", [Line, Col]);
+fmt_pos(Line) ->
+    io_lib:format("~w", [Line]).
+
+quote_source(File, StartLoc, EndLoc, Opts) ->
+    case proplists:get_bool(brief, Opts) of
+        true -> "";
+        false -> quote_source_1(File, StartLoc, EndLoc)
+    end.
+
+quote_source_1(File, Line, Loc2) when is_integer(Line) ->
+    quote_source_1(File, {Line, 1}, Loc2);
+quote_source_1(File, Loc1, Line) when is_integer(Line) ->
+    quote_source_1(File, Loc1, {Line, -1});
+quote_source_1(File, {StartLine, StartCol}, {EndLine, EndCol}) ->
+    case file:read_file(File) of
+        {ok, Bin} ->
+            Ctx =
+                if
+                    StartLine =:= EndLine -> 0;
+                    true -> 1
+                end,
+            case seek_line(Bin, 1, StartLine - Ctx) of
+                {ok, Bin1} ->
+                    quote_source_2(Bin1, StartLine, StartCol, EndLine, EndCol, Ctx);
+                error ->
+                    ""
+            end;
+        {error, _} ->
+            ""
+    end.
+
+quote_source_2(Bin, StartLine, StartCol, EndLine, EndCol, Ctx) ->
+    case take_lines(Bin, StartLine - Ctx, EndLine + Ctx) of
+        [] ->
+            "";
+        Lines ->
+            Lines1 =
+                case length(Lines) =< (4 + Ctx) of
+                    true ->
+                        Lines;
+                    false ->
+                        %% before = context + start line + following line
+                        %% after = end line + context
+                        %% (total lines: 3 + 1 + context)
+                        Before = lists:sublist(Lines, 2 + Ctx),
+                        After = lists:reverse(
+                            lists:sublist(lists:reverse(Lines), 1 + Ctx)
+                        ),
+                        Before ++ [{0, "..."}] ++ After
+                end,
+            Lines2 = decorate(Lines1, StartLine, StartCol, EndLine, EndCol),
+            [[fmt_line(L, Text) || {L, Text} <- Lines2], $\n]
+    end.
+
+line_prefix() ->
+    "% ".
+
+fmt_line(L, Text) ->
+    io_lib:format("~ts~4.ts| ~ts\n", [line_prefix(), line_to_txt(L), Text]).
+
+line_to_txt(0) -> "";
+line_to_txt(L) -> integer_to_list(L).
+
+decorate([{Line, Text} = L | Ls], StartLine, StartCol, EndLine, EndCol) when
+    Line =:= StartLine, EndLine =:= StartLine
+->
```

Comment:
```suggestion
    Line =:= StartLine, EndLine =:= StartLine ->
```

Reply:
Some remnants of running erlfmt on my original patch. Fixing.